tests/subsys/net/lib/download_client/src/main.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <logging/log.h>
+LOG_MODULE_REGISTER(download_client_test, LOG_LEVEL_DBG);
+
+#include <zephyr.h>
+#include "net_private.h"
+#include <net/socket_offload.h>
+
+#include <ztest.h>
+#include <download_client.h>
+
+#include "mock/socket.h"
+#include "mock/dl_coap.h"
+
+static enum download_client_evt_id last_event = -1;
+
+static int download_client_callback(const struct download_client_evt *event)
+{
+	if (event == NULL) {
+		return -EINVAL;
+	}
+
+	switch (event->id) {
+	case DOWNLOAD_CLIENT_EVT_FRAGMENT:
+		last_event = DOWNLOAD_CLIENT_EVT_FRAGMENT;
+		break;
+	case DOWNLOAD_CLIENT_EVT_DONE:
+		last_event = DOWNLOAD_CLIENT_EVT_DONE;
+		break;
+	case DOWNLOAD_CLIENT_EVT_ERROR:
+		last_event = DOWNLOAD_CLIENT_EVT_ERROR;
+		break;
+	}
+
+	return 0;
+}
+
+static enum download_client_evt_id get_last_download_client_event(bool reset)
+{
+	enum download_client_evt_id event;
+
+	event = last_event;
+	if (reset) {
+		last_event = -1;
+	}
+
+	return event;
+}
+
+static void mock_return_values(const char *func, int32_t *val, size_t len)
+{
+	size_t i;
+
+	for (i = 0; i < len; i++) {
+		z_ztest_returns_value(func, val[i]);
+	}
+}
+
+static bool wait_for_event(enum download_client_evt_id event, uint8_t seconds)
+{
+	int i;
+
+	for (i = 0; i < (20 * seconds); i++) {
+		if (get_last_download_client_event(false) == event) {
+			break;
+		}
+
+		k_sleep(K_MSEC(10));
+	}
+
+	return get_last_download_client_event(true) != event;
+}
+
+static struct download_client_cfg config = {
+	.sec_tag = -1,
+	.pdn_id = 0,
+	.frag_size_override = 0,
+};
+
+static void dl_coap_start(struct download_client *client)
+{
+	static const char host[] = "coap://10.1.0.10";
+	int err;
+
+	memset(client, 0, sizeof(struct download_client));
+
+	err = download_client_init(client, download_client_callback);
+	zassert_ok(err, NULL);
+	// FIXME: possible race condition; give time to client_init to start the dl thread
```

Comment:
Maybe it'd be enough to align `ZTEST_THREAD_PRIORITY` with download thread prio, and `k_yield()` here? This should give the download thread a chance to run.

Reply:
actually it is not needed anymore as I just have tested.